src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -54,71 +49,6 @@ class GTPair(val p: Int) extends AnyVal {
 }
 
 
-abstract class Genotype extends Serializable {
-
-  def _unboxedGT: Int
-
-  def _unboxedAD: Array[Int]
-
-  def _unboxedDP: Int
-
-  def _unboxedGQ: Int
-
-  def _unboxedPX: Array[Int]
-
-  def check(nAlleles: Int) {
-    val nGenotypes = triangle(nAlleles)
-    assert(Genotype.gt(this).forall(i => i >= 0 && i < nGenotypes))
-    assert(Genotype.ad(this).forall(a => a.length == nAlleles))
-  }
-
-  def copy(gt: Option[Int] = Genotype.gt(this),
-    ad: Option[Array[Int]] = Genotype.ad(this),
-    dp: Option[Int] = Genotype.dp(this),
-    gq: Option[Int] = Genotype.gq(this),
-    px: Option[Array[Int]] = Genotype.px(this)): Genotype = Genotype(gt, ad, dp, gq, px)
-
-  override def equals(that: Any): Boolean = that match {
-    case g: Genotype =>
-      _unboxedGT == g._unboxedGT &&
-        util.Arrays.equals(_unboxedAD, g._unboxedAD) &&
-        _unboxedDP == g._unboxedDP &&
-        _unboxedGQ == g._unboxedGQ &&
-        util.Arrays.equals(_unboxedPX, g._unboxedPX)
-
-    case _ => false
-  }
-
-  override def hashCode: Int =
-    new HashCodeBuilder(43, 19)
-      .append(_unboxedGT)
-      .append(util.Arrays.hashCode(_unboxedAD))
-      .append(_unboxedDP)
-      .append(_unboxedGQ)
-      .append(util.Arrays.hashCode(_unboxedPX))
-      .toHashCode
-
-  override def toString: String = {
-    val b = new StringBuilder
-
-    b.append(Genotype.gt(this).map { gt =>
-      val p = Genotype.gtPair(gt)
-      s"${ p.j }/${ p.k }"
-    }.getOrElse("./."))
-
-    b += ':'
-    b.append(Genotype.ad(this).map(_.mkString(",")).getOrElse("."))
-    b += ':'
-    b.append(Genotype.dp(this).map(_.toString).getOrElse("."))
-    b += ':'
-    b.append(Genotype.gq(this).map(_.toString).getOrElse("."))
-    b += ':'
-    b.append("PL=" + Genotype.pl(this).map(_.mkString(",")).getOrElse("."))
-
-    b.result()
-  }
-}
-
 object Genotype {
   val htsGenotypeType: TStruct = TStruct(
```

Comment:
This doesn't seem like the long-term design we want, because in order to run variantQC, you don't need something that is exactly GT/AD/DP/GQ/PL, but just something that contains those fields in any order (and could contain others). So we wouldn't want to check equality, but instead to check membership. Is this a temporary workaround to keep stuff working with the tests?

Reply:
This isn't used in variantQC.  It used in split_multi and related routines.

split_multi requires exactly this schema (although in an arbitrary order is fine), because it has a fixed schema on output and will delete any other genotype fields present.  Although the input can be in any order, the output is in the standard order.  I have to do this because we can't put an insertion inside a let and I need, for example, to compute GQ  from PL (without recording PL twice).  Also, instantiating the entire struct is way faster than doing 5 insertions (although that's a short term concern).

So ... I could make split_multi require the right fields but in any order.

Reply:
Alright. Let's revisit this again in a month or so, to evaluate whether it's still the right thing. Seems fine for now.